### PR TITLE
`Development`: Migrate the git diff report module to standalone components

### PR DESCRIPTION
**Pull request by pzdr7**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

The new client coding guidelines specify that we should use signals/standalone components.

### Description

- Migrated all components in the GitDiffReportModule to standalone components
- Introduced signals, effects, and the OnPush change detection in those components
- Simplified the layout handling of the diff editor to make it less error-prone

### Steps for Testing

Prerequisites:
- Test Server with LocalVC (e.g. TS1-TS3)
- 1 Instructor
- 1 Programming Exercise

1. Navigate to Course management > Your course > Exercises and click the title of your programming exercise.
2. Scroll down to "Review changes" and check that the modal opens as expected.
3. Close the modal.
4. Navigate to the repository view of any repository (e.g. for the template repository. Click the "Code" button and 
image to open the repository view).
5. Click "Open commit history" and then on a commit hash
6. Verify that the commit diff is displayed correctly

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| programming-diff-report-detail.component.ts | 100% | ✅ |
| programming-exercise-exam-diff.component.ts | 97.26% | ✅ |
| git-diff-file-panel-title.component.ts | 100% | ✅ |
| git-diff-file-panel.component.ts | 100% | ✅ |
| git-diff-file.component.ts | 100% | ✅ |
| git-diff-line-stat.component.ts | 100% | ✅ |
| git-diff-report-modal.component.ts | 100% | ✅ |
| git-diff-report.component.ts | 98.66% | ✅  |
| monaco-diff-editor.component.ts | 98% | ✅ |
| monaco-editor.service.ts | 100% | ✅ |

### Screenshots

No UI changes

## Release Notes

- **New Features**
  - Introduced a new `GitDiffLineStatComponent` for enhanced line statistics.
  - Added `GitDiffReportComponent` for improved reporting on Git diffs.

- **Improvements**
  - Transitioned several components to utilize Angular's reactive signals for better state management.
  - Updated various components to use method calls for property access, enhancing dynamic behavior.
  - Enhanced the Monaco Diff Editor with new styling and automatic layout configuration.

- **Bug Fixes**
  - Adjusted modal behavior and test cases for better reliability.

- **Chores**
  - Removed deprecated modules and updated styles for the Monaco Diff Editor.

### CODE REVIEW

File: src/main/webapp/app/detail-overview-list/components/programming-diff-report-detail/programming-diff-report-detail.component.ts [hunk 1]
```diff
@@ -39,6 +39,6 @@ export class ProgrammingDiffReportDetailComponent implements OnDestroy {
         }
 
         this.modalRef = this.modalService.open(GitDiffReportModalComponent, { windowClass: 'diff-view-modal' });
-        this.modalRef.componentInstance.report = gitDiff;
+        this.modalRef.componentInstance.report = signal<ProgrammingExerciseGitDiffReport>(gitDiff);
```
Comment (pzdr7) [inline src/main/webapp/app/detail-overview-list/components/programming-diff-report-detail/programming-diff-report-detail.component.ts:42]:
```suggestion
        this.modalRef.componentInstance.report = signal(gitDiff);
```

File: src/main/webapp/app/exam/manage/student-exams/student-exam-timeline/programming-exam-diff/programming-exercise-exam-diff.component.ts [hunk 2]
```diff
@@ -124,9 +124,9 @@ export class ProgrammingExerciseExamDiffComponent extends ExamPageComponent impl
      */
     showGitDiff(): void {
         const modalRef = this.modalService.open(GitDiffReportModalComponent, { size: 'xl' });
```
Comment (pzdr7) [inline src/main/webapp/app/exam/manage/student-exams/student-exam-timeline/programming-exam-diff/programming-exercise-exam-diff.component.ts:126]:
```suggestion
        const modalRef = this.modalService.open(GitDiffReportModalComponent, { windowClass: 'diff-view-modal' });
```

File: src/main/webapp/app/exam/manage/student-exams/student-exam-timeline/programming-exam-diff/programming-exercise-exam-diff.component.ts [hunk 3]
```diff
@@ -123,10 +123,10 @@ export class ProgrammingExerciseExamDiffComponent extends ExamPageComponent impl
...
      */
     showGitDiff(): void {
-        const modalRef = this.modalService.open(GitDiffReportModalComponent, { size: 'xl' });
-        modalRef.componentInstance.report = this.exercise.gitDiffReport;
-        modalRef.componentInstance.diffForTemplateAndSolution = false;
-        modalRef.componentInstance.cachedRepositoryFiles = this.cachedRepositoryFiles;
+        const modalRef = this.modalService.open(GitDiffReportModalComponent, { windowClass: GitDiffReportModalComponent.WINDOW_CLASS });
+        modalRef.componentInstance.report = signal(this.exercise.gitDiffReport);
+        modalRef.componentInstance.diffForTemplateAndSolution = signal(false);
+        modalRef.componentInstance.cachedRepositoryFiles = signal(this.cachedRepositoryFiles);
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exam/manage/student-exams/student-exam-timeline/programming-exam-diff/programming-exercise-exam-diff.component.ts:129]: This is currently the only way until NgbModal supports signal inputs

File: src/test/javascript/spec/component/hestia/git-diff-report/git-diff-modal.component.spec.ts [hunk 4]
```diff
@@ -52,68 +52,80 @@ describe('GitDiffReportModalComponent', () => {
         jest.restoreAllMocks();
     });
 
-    it('should call correct service method onInit', fakeAsync(() => {
+    const finishEffects = async () => {
+        // We need to wait for all asynchronous code to finish execution before we can expect the results.
+        // See https://stackoverflow.com/questions/44741102/how-to-make-jest-wait-for-all-asynchronous-code-to-finish-execution-before-expec/51045733#51045733
+        await new Promise(process.nextTick);
+        await fixture.whenStable();
+    };
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/test/javascript/spec/component/hestia/git-diff-report/git-diff-modal.component.spec.ts:60]: This doesn't work here

File: src/test/javascript/spec/component/hestia/git-diff-report/git-diff-file.component.spec.ts [hunk 5]
```diff
@@ -33,28 +33,21 @@
...
         const diffEntry = getDiffEntryWithPaths(fileName, fileName);
-        comp.originalFileContent = originalContent;
-        comp.modifiedFileContent = modifiedContent;
-        comp.diffEntries = [diffEntry];
+        fixture.componentRef.setInput('diffEntries', [diffEntry]);
+        fixture.componentRef.setInput('originalFileContent', originalContent);
+        fixture.componentRef.setInput('originalFilePath', fileName);
+        fixture.componentRef.setInput('modifiedFileContent', modifiedContent);
+        fixture.componentRef.setInput('modifiedFilePath', fileName);
+        jest.spyOn(comp.monacoDiffEditor(), 'setFileContents').mockImplementation();
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/test/javascript/spec/component/hestia/git-diff-report/git-diff-file.component.spec.ts:48]: I will remove this

**Review by JohannesWt — CHANGES_REQUESTED** [review]
Code looks good just two small ideas for simplicity

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts [hunk 6]
```diff
@@ -67,65 +43,97 @@ export class GitDiffReportComponent implements OnInit {
...
+    readonly diffInformationForPaths = signal<DiffInformation[]>([]);
+    readonly nothingToDisplay = computed(() => this.diffInformationForPaths().length === 0);
+    readonly allDiffsReady = computed(() => Object.values(this.diffInformationForPaths()).every((info) => info.diffReady));
+    readonly allowSplitView = signal<boolean>(true);
+
+    protected readonly faSpinner = faSpinner;
+    protected readonly faTableColumns = faTableColumns;
+    protected readonly ButtonSize = ButtonSize;
+    protected readonly ButtonType = ButtonType;
+    protected readonly TooltipPlacement = TooltipPlacement;
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts]: [Moved ](https://github.com/ls1intum/Artemis/pull/9443/commits/6601bfccb6fd1dcc74d01efd9016c1f48805ca04)

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report-modal.component.ts [hunk 7]
```diff
@@ -1,151 +1,152 @@
...
+    constructor() {
+        effect(
+            async () => {
+                // We call the signal here to ensure the effect always runs when the report changes.
+                this.report();
+                if (this.diffForTemplateAndSolution()) {
+                    untracked(async () => await this.loadFilesForTemplateAndSolution());
+                } else {
+                    untracked(async () => await this.loadRepositoryFilesForParticipationsFromCacheIfAvailable());
+                }
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report-modal.component.ts]: [Changed](https://github.com/ls1intum/Artemis/pull/9443/commits/e8bf525a326b0d92db3899b1c8f88ae08d062367)

**Review by JohannesWt — APPROVED** [review]
Thanks for the changes - Code 👍

**Review by edkaya — APPROVED** [review]
changes lgtm

**Review by krusche — APPROVED** [review]

**Pull request merged**